klippy/extras/dockable_probe.py

```diff
@@ -0,0 +1,634 @@
+
+    # Determine the vector of two points
+    def _get_vector(self, point1, point2):
+        x1, y1 = point1[:2]
+        x2, y2 = point2[:2]
+        magnitude = sqrt((x2-x1)**2 + (y2-y1)**2 )
+        angle = atan2(y2-y1, x2-x1) + pi
+
+        return angle, magnitude
+
+    # Align z axis to prevent crashes
+    def _align_z(self):
+        curtime = self.printer.get_reactor().monotonic()
+        homed_axes = self.toolhead.get_status(curtime)['homed_axes']
+        self._last_homed = homed_axes
+
+        if self.dock_requires_z:
+            self._align_z_required()
+
+        if self.z_hop > 0.0:
+            if 'z' in self._last_homed:
+                tpos = self.toolhead.get_position()
+                if tpos[2] < self.z_hop:
+                    self.toolhead.manual_move([None, None, self.z_hop],
+                        self.lift_speed)
+            else:
+                self._force_z_hop()
+
+    def _align_z_required(self):
+        if 'z' not in self._last_homed:
+            raise self.printer.command_error(
+                "Cannot attach/detach probe, must home Z axis first")
+
+        self.toolhead.manual_move([None, None, self.approach_position[2]],
+                                self.lift_speed)
+
+    # Hop z and return to un-homed state
+    def _force_z_hop(self):
+        this_z = self.toolhead.get_position()[2]
+        if self.last_z == this_z:
+            return
+
+        tpos = self.toolhead.get_position()
+        self.toolhead.set_position([tpos[0], tpos[1], 0.0, tpos[3]],
+                            homing_axes=[2])
+        self.toolhead.manual_move([None, None, self.z_hop],
+            self.lift_speed)
+        kin = self.toolhead.get_kinematics()
+        kin.note_z_not_homed()
+        self.last_z = self.toolhead.get_position()[2]
+
+    #######################################################################
+    # Probe Wrappers
+    #######################################################################
+
+    def multi_probe_begin(self):
+        self.multi = MULTI_FIRST
+
+        # Attach probe before moving to the first probe point and
+        # return to current position. Move because this can be called
+        # before a multi _point_ probe and a multi probe at the same
+        # point but for the latter the toolhead is already in position.
+        # If the toolhead is not returned to the current position it
+        # will complete the probing next to the dock.
+        return_pos = self.toolhead.get_position()
+        self.auto_attach_probe(return_pos)
+
+    def multi_probe_end(self):
+        self.multi = MULTI_OFF
+
+        return_pos = self.toolhead.get_position()
+        # Move away from the bed to ensure the probe isn't triggered,
+        # preventing detaching in the event there's no probe/dock sensor.
+        self.toolhead.manual_move([None, None, return_pos[2]+2],
+                                  self.travel_speed)
+        self.auto_detach_probe(return_pos)
+
```

Comment:
please add for fix problem after last update https://github.com/Klipper3d/klipper/commit/acdf8bb1080ab829c6e8092dadf6d488aa416f77

`
def probing_move(self, pos, speed):
        phoming = self.printer.lookup_object('homing')
        return phoming.probing_move(self, pos, speed)
`